Add MatchAny to Tags

alexfca opened the pull request:
# Add MatchAny to Tags

## Description
Add the ability to use MatchAny as a LINQ extension method.
Uses a similar approach to Match with the ability to combine ANDs/ORs for Tags

eg.
`query.Where(doc => MatchAny(new [] {new MatchObjectList(new [] {new MatchObject() {DataTags = doc.TagsField, QueryTags = DisplayClass.matchFilter1.get_Item(0), QueryOptions = Basic, UdfName = "UdfName1"}, new MatchObject() {DataTags = doc.TagsField1, QueryTags = DisplayClass.matchFilter1.get_Item(1), QueryOptions = Basic}}), new MatchObjectList(new [] {new MatchObject() {DataTags = doc.TagsField, QueryTags = DisplayClass.matchFilter2.get_Item(0), QueryOptions = Basic, UdfName = "UdfName1"}, new MatchObject() {DataTags = doc.TagsField1, QueryTags = DisplayClass.matchFilter2.get_Item(1), QueryOptions = Basic}})}))`

will produce
`SELECT VALUE root 
FROM root 
WHERE ((((((NOT(IS_DEFINED(root["TagsField"]["tags"]["ns:tagname"])) OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=") OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=tagvalue1") OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=tagvalue2")))) AND (((NOT(IS_DEFINED(root["TagsField1"]["tags"]["ns:tagname1"])) OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=") OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=tag1value1") OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=tag1value2")))))) OR (((((NOT(IS_DEFINED(root["TagsField"]["tags"]["ns:tagname"])) OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=") OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=tagvalue3") OR ARRAY_CONTAINS(root["TagsField"]["tag"], "ns:tagname=tagvalue4")))) AND (((NOT(IS_DEFINED(root["TagsField1"]["tags"]["ns:tagname1"])) OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=") OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=tag1value3") OR ARRAY_CONTAINS(root["TagsField1"]["tag"], "ns:tagname1=tag1value4")))))))`

Add Tags MatchAny clause to support generating AND/OR expressions for Tag matching
- Add new MatchAny to the CosmosTags.cs that supports using a MatchObjectList
- Add the MatchAny to the ExpressionToSQL.cs and include LINQ handling of specific Select/Array
- Implement the AND/OR handling as part of the Visit() in the SqlObjectTextSerializer.cs
- Update the relevant classes to implement the new expressions
- Update the Unit Test generation using UpdateContracts.ps1
- Increment the NuGet package version to 3.39.1001-ca

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)
To automatically close an issue: closes #IssueNumber

alexfca committed:
Add Tags MatchAny clause to support generating AND/OR expressions for Tag matching
- Add new MatchAny to the CosmosTags.cs that supports using a MatchObjectList
- Add the MatchAny to the ExpressionToSQL.cs and include LINQ handling of specific Select/Array
- Implement the AND/OR handling as part of the Visit() in the SqlObjectTextSerializer.cs
- Update the relevant classes to implement the new expressions
- Update the Unit Test generation using UpdateContracts.ps1
- Increment the NuGet package version to 3.39.1001-ca

jahmai-ca commented:
QQ: I thought you made a change that shortened the MatchObject syntax to avoid using all the property setters?

alexfca commented:
> QQ: I thought you made a change that shortened the MatchObject syntax to avoid using all the property setters?

Sorry, yeah must have missed that. will double check and add back in, sorry

alexfca committed:
Add extra constructors for the MatchObject
- Add MatchObject() and MatchObject.Create() to the CosmosTags.cs
- Update the ExpressionToSQL.cs to handle the new ctors and simplify the expressions
- Add a new UdfNameDefault to the CosmosTags.cs
- Add tests to the LinqTranslationBaselineTests.cs and update the LinqTranslationBaselineTests.TestSpecialMethods.xml

alexfca commented:
> > QQ: I thought you made a change that shortened the MatchObject syntax to avoid using all the property setters?
> 
> Sorry, yeah must have missed that. will double check and add back in, sorry

Ok all done, forgot to include that latest set of changes.
All still works and makes the usage nicer, thanks for that

alexfca committed:
Optimize for single set of filters to reduce the number of brackets
- Update the SqlObjectTextSerializer.cs to check the item count. When 1 just Visit, else Loop and Visit
- Update tests
